File: runners/kre-go/context_prediction.go
package kre

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/konstellation-io/kre/libs/simplelogger"
	"github.com/nats-io/nats.go"

	"github.com/konstellation-io/kre/runners/kre-go/config"
)

type SaveMetricErr string

const (
	classificationMetricsColl               = "classificationMetrics"
	ErrMissingValues          SaveMetricErr = "missing_values"
	ErrNewLabels              SaveMetricErr = "new_labels"
)

func (s SaveMetricErr) IsValid() error {
	switch s {
	case ErrMissingValues, ErrNewLabels:
		return nil
	}
	return fmt.Errorf("invalid SaveMetricErr type: %s", s)
}

type SaveMetricMsg struct {
	Coll string           `json:"coll"`
	Doc  SaveMetricMsgDoc `json:"doc"`
}

type SaveMetricMsgDoc struct {
	Date           string `json:"date"`
	Error          string `json:"error"`
	PredictedValue string `json:"predictedValue"`
	TrueValue      string `json:"trueValue"`
	VersionID      string `json:"versionId"`
	VersionName    string `json:"versionName"`
}

type contextPrediction struct {
	cfg    config.Config
	nc     *nats.Conn
	logger *simplelogger.SimpleLogger
}

func (c *contextPrediction) Save(date time.Time, predictedValue, trueValue string) {
	msg, err := json.Marshal(SaveMetricMsg{
		Coll: classificationMetricsColl,
		Doc: SaveMetricMsgDoc{
			Date:           date.Format(time.RFC3339),
			PredictedValue: predictedValue,
			TrueValue:      trueValue,
			VersionID:      c.cfg.VersionID,
			VersionName:    c.cfg.Version,
		},
	})

	if err != nil {
		c.logger.Infof("Error marshalling SaveMetricMsgDoc: %s", err)
		return
	}

	_, err = c.nc.Request(c.cfg.NATS.MongoWriterSubject, msg, saveMetricTimeout)
	if err != nil {
		c.logger.Infof("Error sending metric to NATS: %s", err)
	}
}

func (c *contextPrediction) SaveError(saveMetricErr SaveMetricErr) {
	if err := saveMetricErr.IsValid(); err != nil {
		c.logger.Error(err.Error())
		return
	}

	msg, err := json.Marshal(SaveMetricMsg{
		Coll: classificationMetricsColl,
		Doc: SaveMetricMsgDoc{
			Date:        time.Now().Format(time.RFC3339),
			VersionName: c.cfg.Version,
			Error:       string(saveMetricErr),
		},
	})
	if err != nil {
		c.logger.Infof("Error generating SaveMetricMsg JSON: %s", err)
	}

	_, err = c.nc.Request(c.cfg.NATS.MongoWriterSubject, msg, saveMetricTimeout)
	if err != nil {
		c.logger.Infof("Error sending error metric to NATS: %s", err)
	}
}

File: runners/kre-py/src/context_prediction.py
import json
from datetime import datetime

from nats.aio.client import ErrTimeout


class ContextPrediction:
    ERR_MISSING_VALUES = "missing_values"
    ERR_NEW_LABELS = "new_labels"

    def __init__(self, config, nc, logger):
        self.PREDICTION_ERRORS = [self.ERR_MISSING_VALUES, self.ERR_NEW_LABELS]
        self.__config__ = config
        self.__nc__ = nc
        self.__logger__ = logger

    async def save(self, predicted_value: str = "", true_value: str = "", utcdate: datetime = None,
                   error: str = "") -> None:
        if error != "":
            if error not in self.PREDICTION_ERRORS:
                raise Exception(f"[ctx.prediction.save] invalid value for metric error {error} "
                                f"should be one of '{','.join(self.PREDICTION_ERRORS)}'")
        else:
            if not isinstance(predicted_value, str) or predicted_value == "":
                raise Exception(f"[ctx.prediction.save] invalid 'predicted_value'='{predicted_value}',"
                                f" must be a nonempty string")
            if not isinstance(true_value, str) or true_value == "":
                raise Exception(f"[ctx.prediction.save] invalid 'true_value'='{true_value}', must be a nonempty string")

        if utcdate is None:
            utcdate = datetime.utcnow()

        coll = "classificationMetrics"
        doc = {
            "date": utcdate_to_rfc3339(utcdate),
            "error": error,
            "predictedValue": predicted_value,
            "trueValue": true_value,
            "versionId": self.__config__.krt_version_id,
            "versionName": self.__config__.krt_version
        }

        try:
            subject = self.__config__.nats_mongo_writer
            payload = bytes(json.dumps({"coll": coll, "doc": doc}), encoding='utf-8')
            response = await self.__nc__.request(subject, payload, timeout=1)
            res_json = json.loads(response.data.decode())
            if not res_json['success']:
                self.__logger__.error("Unexpected error saving metric")
        except ErrTimeout:
            self.__logger__.error("Error saving metric: request timed out")


def utcdate_to_rfc3339(utcdate):
    offset = utcdate.utcoffset()
    if offset and offset.total_seconds() != 0:
        raise Exception("Input date must be a UTC datetime")

    return utcdate.strftime('%Y-%m-%dT%H:%M:%S.%fZ')
